src/blockchains/icp/icp_worker.ts

```diff
@@ -203,9 +218,25 @@ export class ICPWorker extends BaseWorker {
     const requests = Array.from({ length: numConcurrentRequests }, (_, i) => this.fetchBlock(this.lastExportedBlock + 1 + i));
     const blocks = await Promise.all(requests);
 
-    const transactions = await this.getTransactions(blocks);
-    
+    let transactions: (Transaction)[] = await this.getTransactions(blocks);
+    let extendedTransactions: (ExtendedTransaction)[] = [];
+    if (transactions.length > 0) {
+      stableSort(transactions, transactionOrder);
+      extendedTransactions = extendTransactionsWithPrimaryKey(transactions, this.lastPrimaryKey);
+
+      this.lastPrimaryKey += transactions.length;
+    }
+
     this.lastExportedBlock += blocks.length;
-    return transactions;
+    return extendedTransactions;
   }
 }
+
+export function extendTransactionsWithPrimaryKey(transactions: Transaction[], lastPrimaryKey: number): ExtendedTransaction[] {
```

Comment:
This approach of setting primary key is not ideal. We use a global incrementing counter, this causes problems when we re-extract the data  and we see different order of events or different number of events. In this case primary keys between first and second re-extraction would change and we can no longer treat the records as the same in CH. Ideal case is to use some on-chain characteristics that the blockchain protocol gives us. Like event position inside transaction. Not sure if ICP has this, just a food for thought.

Reply:
There is indeed transaction index, which we could use to have deterministic primary keys. Let's merge this PR and I'll use it in icrc(icp tokens) exporter and after it is done, update icp to new version which will use them. wdys?